packages/@n8n/task-runner/src/task-runner.ts

```diff
@@ -284,111 +286,116 @@ export abstract class TaskRunner extends EventEmitter {
+			},
+		});
+	}
+
+	async taskTimedOut(taskId: string) {
+		const taskState = this.runningTasks.get(taskId);
+		if (!taskState) {
+			return;
 		}
 
-		if (!task.active) this.runningTasks.delete(taskId);
+		await taskState.caseOf({
+			// If we are still waiting for settings for the task, we can error the
+			// task immediately
+			waitingForSettings: () => {
+				this.send({
+					type: 'runner:taskerror',
+					taskId,
+					error: new TimeoutError(this.taskTimeout),
+				});
 
-		this.sendOffers();
-	}
+				this.finishTask(taskState);
+			},
 
-	taskErrored(taskId: string, error: unknown) {
-		this.send({
-			type: 'runner:taskerror',
-			taskId,
-			error,
-		});
-		this.runningTasks.delete(taskId);
-		this.sendOffers();
-	}
+			// This should never happen, the time out timer should only fire once
+			'aborting:timeout': TaskState.throwUnexpectedTaskStatus,
 
-	taskDone(taskId: string, data: RunnerMessage.ToBroker.TaskDone['data']) {
-		this.send({
-			type: 'runner:taskdone',
-			taskId,
-			data,
+			// If we are currently executing the task, abort the execution and
+			// mark the task as timed out
+			running: () => {
+				taskState.status = 'aborting:timeout';
+				taskState.abortController.abort('timeout');
+				this.cancelTaskRequests(taskId, 'timeout');
+			},
+
+			// If the task is already cancelling, we can ignore the timeout
+			'aborting:cancelled': noOp,
 		});
-		this.runningTasks.delete(taskId);
-		this.sendOffers();
 	}
 
 	async receivedSettings(taskId: string, settings: unknown) {
-		const task = this.runningTasks.get(taskId);
-		if (!task) {
-			return;
-		}
-		if (task.cancelled) {
-			this.runningTasks.delete(taskId);
+		const taskState = this.runningTasks.get(taskId);
+		if (!taskState) {
 			return;
 		}
 
-		const controller = new AbortController();
-		this.taskCancellations.set(taskId, controller);
-
-		const taskTimeout = setTimeout(() => {
-			if (!task.cancelled) {
-				controller.abort();
-				this.taskCancellations.delete(taskId);
-			}
-		}, this.taskTimeout * 1_000);
-
-		task.settings = settings;
-		task.active = true;
-		try {
-			const data = await this.executeTask(task, controller.signal);
-			this.taskDone(taskId, data);
-		} catch (error) {
-			if (!task.cancelled) this.taskErrored(taskId, error);
-		} finally {
-			clearTimeout(taskTimeout);
-			this.taskCancellations.delete(taskId);
-			this.resetIdleTimer();
-		}
+		await taskState.caseOf({
+			// These states should never happen, as they are handled already in
+			// the other lifecycle methods and the task should be removed from the
+			// running tasks
+			'aborting:cancelled': TaskState.throwUnexpectedTaskStatus,
+			'aborting:timeout': TaskState.throwUnexpectedTaskStatus,
+			running: TaskState.throwUnexpectedTaskStatus,
+
+			waitingForSettings: async () => {
+				taskState.status = 'running';
+
+				await this.executeTask(
+					{
+						taskId,
+						settings,
+					},
+					taskState.abortController.signal,
+				)
```

Comment:
Transitions aren't atomic - what if a cancellation arrives at L378? Too unlikely?

Reply:
That is not possible, since Nodejs is single threaded